nntrainer/layers/pooling2d_layer.cpp

```diff
@@ -98,12 +115,28 @@ void Pooling2DLayer::calcDerivative() {
   Tensor &result = net_input[0]->getGradientRef();
 
   result.setZero();
-  float *out = result.getData();
+  float *result_data = result.getData();
+
+  unsigned int out_map_size = deriv.height() * deriv.width();
+  unsigned int in_map_size = height * width;
 
   switch (pooling_type) {
   case PoolingType::max: {
-    for (unsigned int i = 0; i < deriv.getDim().getDataLen(); ++i) {
-      out[max_idx[i]] += deriv.getData()[i];
+    auto iter = max_idx.begin();
+    const float *deriv_data = deriv.getData();
+    for (unsigned int b = 0; b < batch; ++b) {
+      for (unsigned int c = 0; c < channel; ++c) {
+        for (unsigned int i = 0; i < out_map_size; ++i) {
+          /// max_idx = -1 means the max idx was at the padding, so no need to
+          /// update
+          if (*iter != -1) {
```

Comment:
as `iter == -1` is ignored, is there any reason to store `-1` values in `max_idx` than rather not storing these indices in it at all?

Reply:
max_idx == -1 has information that for this patch, calcDerivative has to be skipped. if we don't keep this, the memory layout breaks